src/stratigraphy/groundwater/groundwater_extraction.py

```diff
@@ -0,0 +1,420 @@
+
+        Args:
+            lines (list[TextLine]): all the lines of text to search in
+            page (int): the page number (1-based) of the PDF document
+            page_width (float): the width of the current page (in points / PyMuPDF coordinates)
+
+        Returns:
+            list[Coordinate]: all found groundwater information
+        """
+        # find the key that indicates the groundwater information
+        groundwater_key_lines = self.find_groundwater_key(lines)
+        if groundwater_key_lines is None:
+            return []
+
+        extracted_groundwater_informations = []
+
+        for groundwater_key_line in groundwater_key_lines:
+            # find the lines of the text that are close to an identified groundwater key.
+            key_rect = groundwater_key_line.rect
+            # look for groundwater related values to the right and/or immediately below the key
+            groundwater_info_search_rect = fitz.Rect(
+                max(0, key_rect.x0 - 2.0 * key_rect.width), key_rect.y0, page_width, key_rect.y1 + 3 * key_rect.height
+            )
+            groundwater_info_lines = [line for line in lines if line.rect.intersects(groundwater_info_search_rect)]
+
+            def preprocess(value: str) -> str:
+                value = value.replace(",", ".")
+                value = value.replace("'", ".")
+                value = value.replace("o", "0")
+                value = value.replace("\n", " ")
+                value = value.replace("ü", "u")
+                return value
+
+            # makes sure the line with the key is included first in the extracted information and the duplicate removed
+            groundwater_info_lines.insert(0, groundwater_key_line)
+            groundwater_info_lines = list(dict.fromkeys(groundwater_info_lines))
+
+            # sort the lines by their proximity to the key line center, compute the distance to the key line center
+            key_center = (key_rect.x0 + key_rect.x1) / 2
+            groundwater_info_lines.sort(key=lambda line: abs((line.rect.x0 + line.rect.x1) / 2 - key_center))
+
+            try:
+                extracted_gw_information = self.get_groundwater_info_from_lines(
+                    groundwater_info_lines, page, preprocess
+                )
+                if extracted_gw_information.depth:
+                    extracted_groundwater_informations.append(extracted_gw_information)
+            except ValueError:
+                logger.warning("Could not extract groundwater information from the lines near the key.")
+
+        return self.select_best_groundwater_information(extracted_groundwater_informations)
+
+    def select_best_groundwater_information(
+        self, extracted_groundwater_informations: list[GroundwaterInformation]
+    ) -> GroundwaterInformation | None:
+        """Selects the best groundwater information from a list of extracted groundwater information.
```

Comment:
Why do this, instead of returning all extracted groundwater information?
The Zürich ground truth does contain at most one entry per borehole profile, but that is a limitation of the ground truth, not something that we actually want to imitate when extracting ground water data. Here, we do want to extract ALL available ground water level measurements, if possible.

Reply:
Ah I see. I was trying always to return the latest groundwater measurement that was performed or the most complete (elevation, depth, and date). I observed that by returning the "best" / most complete, the accuracy also improved since, in general, for a false positive keyword match, it would be hard to find elevation/depth/date around the keyword.